openedx_external_enrollments/tests/external_enrollments/tests_pathstream_external_enrollments.py

```diff
@@ -29,7 +33,332 @@ def test_str(self):
         PathstreamExternalEnrollment overrides the __str__ method,
         this test that the method __str__ returns the right value.
         """
-        self.assertEqual(self.base.__str__(), 'pathstream')
+        self.assertEqual(
+            'pathstream',
+            self.base.__str__(),
+        )
+
+    @patch('openedx_external_enrollments.external_enrollments.pathstream_external_enrollment.boto3', spec=boto3)
+    def test_init_s3_all_settings_setup(self, boto3_mock):
+        """
+        _init_s3 must be called with all parameters
+        """
+        access_key_id = 'access_key'
+        secret_access_key = 'secret_access_key'
+
+        self.base._init_s3()  # pylint: disable=protected-access
+
+        boto3_mock.client.assert_called_with(
+            's3',
+            aws_access_key_id=access_key_id,
+            aws_secret_access_key=secret_access_key,
+        )
+
+    def test_download_file(self):
+        """
+        _download_file must return the downloaded content from the S3 file if called with all parameters and the
+        s3 connection is established.
+        """
+        s3_file = 'test.log'
+        s3_bucket = 'test'
+        expected_file_content = 'file_content_test'
+        self.base.client = Mock(spec=boto3.client('s3'))
+        self.base.client.get_object.return_value = {
+            'Body': io.StringIO(expected_file_content),
+            'ResponseMetadata': {
+                'HTTPStatusCode': 200,
+            },
+        }
+        log = 'File content [{}] was downloaded from S3 for [{}]'.format(s3_file, self.base.__str__())
+
+        with LogCapture(level=logging.INFO) as log_capture:
+            file_content = self.base._download_file()  # pylint: disable=protected-access
+
+            log_capture.check(
+                (self.module, 'INFO', log),
+            )
+
+        self.base.client.get_object.assert_called_with(
+            Bucket=s3_bucket,
+            Key=s3_file,
+        )
+        self.assertEqual(file_content, expected_file_content)
+
+    def test_download_file_error(self):
```

Comment:
Why not self.client.get_object.side_effect = ClientError? Because the error actually occurs on self.base.client, not in _download_file as it is now... @Jacatove

Reply:
The error does occur on `self.base.client` but it raises when `self.base.client.get_object` is called, I set the attributes though keywords arguments (**attrs). However, Now I see that using `self.client.get_object.side_effect = ClientError` is clearer. @Squirrel18